### ISSUE TITLE
[java] BeanMembersShouldSerializeRule does not recognize lombok accessors

### ISSUE THREAD

**Issue by emersonmoura** [pmd/pmd#780]
**Rule Set:**
<ruleref="rulesets/java/design.xml"/>

**Description:**
BeanMembersShouldSerializeRule Does not recognize methods generated by lombok

**Code Sample demonstrating the issue:**

```
import lombok.Data;
@Data
public class Buzz {
    private String foo;
}

```
Lombok will create getters and setters, but "visit" method add
"Found non-transient, non-static member. Please mark as transient or provide accessors"

**Running PMD through:** *[Sonar plugin]*

**Comment by jsotuyod** [comment]
@emersonmoura thanks for the report, certainly this class needs to be lombok-aware.

**Comment by emersonmoura** [comment]
@jsotuyod Because it's like "false-positive"?

**Comment by jsotuyod** [comment]
@emersonmoura I'm sorry, but are you asking why I flagged the issue as a false positive?

It's for the team's triage. We have basically 4 types of issues (there are a couple more, but not so important):
1. requests for new rules / languages (labeled `a:new-language` and `a:new-rule`)
2. requests for enhancement to existing features (labeled `an:enhancement`)
3. bugs / crashes (labeled `a:bug`)
4. rules failing at detecting violations

On this last item, there are 2 possible subtypes: `a:false-positive` and `a:false-negative`, and this has to do with the type of error on the rule.

A false positive is when the rule reports a violation, but it shouldn't, as the code is actually correct (as is this case, since lombok will produce the needed getters / setters). On the contrary, a false negative is when the rule fails to report a violation on code that is *not* correct.

You can read more on false positives and negatives on Wikipedia: https://en.wikipedia.org/wiki/False_positives_and_false_negatives

**Comment by emersonmoura** [comment]
Ah yes!
The label is for the type of error and not for my report.
Thanks.

**Comment by nbali** [comment]
Any progress/ETA/etc on the issue?

**Comment by jsotuyod** [comment]
@nbali I fear not. The issue remains unscheduled. However, fixing this should be far easier now thanks to the introduction of `AbstractIgnoredAnnotationRule` in PMD 6.2.0. As such, I'm marking this issue as a good first issue to encourage anyone willing to help with a PR.

You can also upvote issues that matter most to you to help us prioritize issues for upcoming releases.

**Comment by nbali** [comment]
Thanks anyway.

**Comment by Shubham-2k17** [comment]
I am not able to replicate the issue in current version of PMD. 
Is the issue already resolved?
I Tried adding test-code to BeanMembersShouldSerialize.xml to replicate the issue.
```
 <test-code>
        <description>#780 [java] BeanMembersShouldSerializeRule does not recognize lombok accessors</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
import lombok.Data;
@Data
public class Buzz {
    private String foo;
}
        ]]></code>
</test-code>
```

Few things i noticed. No idea why by following test-code below is working fine and not showing "**Found non-transient, non-static member. Please mark as transient or provide accessors**".
```
<test-code>
        <description>This test code is not showing Found non-transient, non-static member. Please mark as transient or provide accessors</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
public class Buzz {
    private String foo;
}
        ]]></code>
</test-code>
```

But this test-code is showing error.
```
<test-code>
        <description>This test code is showing Found non-transient, non-static member. Please mark as transient or provide accessors.</description>
        <expected-problems>1</expected-problems>
        <code><![CDATA[
public class Buzz {
    private String foo;
    private String bar = foo;
}
        ]]></code>
</test-code>
```
More information to replicate the issue will be helpful.
I would also like to understand how BeanMembersShouldSerializeRule is working in the cases mentioned above.

**Comment by jsotuyod** [comment]
@Shubham-2k17 the code actively filters any field that is not used. This is done in the first clause of this if:

https://github.com/pmd/pmd/blob/dab1fb24c6fe982d0761d3daebecd4d223697c15/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/BeanMembersShouldSerializeRule.java#L84

This is done, since if a field is not used at all, we don't really care for it to be serializable, as it won't be used on the other end either; and is the reason why you see so many test cases with 2 variables (lousy tests, non the less, we should probably tidy those up at some point).

The minimal code snippet to reproduce the issue is therefore a little more elaborate than stated in the issue report:

```java
import lombok.Data;

@Data
public class Buzz {
    private String foo;

    public Buzz(String s) {
        foo = s;
    }
}
```
